src/plugins/content_management/server/event_stream/memory/memory_event_stream_client.ts

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type {
+  EventStreamClient,
+  EventStreamClientFilterOptions,
+  EventStreamClientFilterResult,
+  EventStreamEvent,
+} from '../types';
+import { clone } from './util';
+
+/**
+ * This is an in-memory implementation of the {@link EventStreamClient}
+ * interface (it does not persist events to Elasticsearch). It is useful for
+ * testing and demo purposes.
+ */
+export class MemoryEventStreamClient implements EventStreamClient {
+  #events: EventStreamEvent[] = [];
+
+  public async initialize(): Promise<void> {}
+
+  public async writeEvents(events: EventStreamEvent[]): Promise<void> {
+    for (const event of events) {
+      this.#events.push(clone(event));
+    }
+    this.#events.sort((a, b) => b.time - a.time);
+  }
+
+  public async tail(limit: number = 100): Promise<EventStreamEvent[]> {
+    const tail = this.#events.slice(0, limit);
+
+    return tail.map(clone);
+  }
+
+  public async filter(
+    options: EventStreamClientFilterOptions
+  ): Promise<EventStreamClientFilterResult> {
+    let events: EventStreamEvent[] = [...this.#events];
+
+    const { subject, object, predicate, transaction, from, to } = options;
+
+    if (subject && subject.length) {
+      events = events.filter((event) => {
+        if (!event.subject) {
+          return false;
+        }
+
+        return subject.some(([type, id]) => {
+          if (!id) return type === event.subject![0];
+          return type === event.subject![0] && id === event.subject![1];
+        });
```

Comment:
Depends how you define "needed". In theory, you are right, it was already checked before. But TypeScript shows an error:

<img width="392" alt="image" src="https://user-images.githubusercontent.com/82822460/228560174-3fa55a23-8180-4d64-9eed-90f5921ba1b3.png">

The reason why TypeScript shows an error, is because that check happened in a parent function closure. This one is inside:

```ts
subject.some( () => { ...
```

I theory, TypeScript is right, that `subject.some()` function could already have performed some side-effect, such that `event.subject` is already unavailable, but as developers we know that `subject.some` does not do side-effects.

Reply:
Ah, you're right! It looks like I already sold the idea of `const [subjectType, subjectId] = event.subject;` before the loop to myself :slightly_smiling_face: